tls/s2n_client_cert.c

```diff
@@ -27,6 +27,71 @@
 #include "utils/s2n_blob.h"
 #include "utils/s2n_safety.h"
 
+/* In TLS1.2, the certificate list is just an opaque vector of certificates:
+ *
+ *      opaque ASN.1Cert<1..2^24-1>;
+ *
+ *      struct {
+ *          ASN.1Cert certificate_list<0..2^24-1>;
+ *      } Certificate;
+ *
+ * This construction allowed us to store the entire certificate_list blob
+ * and return it from the s2n_connection_get_client_cert_chain method for
+ * customers to examine.
+ *
+ * However, TLS1.3 introduced per-certificate extensions:
+ *
+ *      struct {
+ *          opaque cert_data<1..2^24-1>;
+ * ---->    Extension extensions<0..2^16-1>;    <----
+ *      } CertificateEntry;
+ *
+ *      struct {
+ *          opaque certificate_request_context<0..2^8-1>;
+ *          CertificateEntry certificate_list<0..2^24-1>;
+ *      } Certificate;
+ *
+ * So in order to store / return the certificates in the same format as in TLS1.2,
+ * we need to first strip out the extensions.
+ */
+static S2N_RESULT s2n_client_cert_chain_store(struct s2n_connection *conn, struct s2n_blob *client_cert_chain)
```

Comment:
won't this difference between TLS1.2 and TLS1.3  impact server cert chain as well?

Reply:
It would if we saved the server cert chain in the same way, but we only save the client cert chain like this. We just validate and then discard the server cert: https://github.com/aws/s2n-tls/blob/main/tls/s2n_server_cert.c#L26